Find the strongest AI

bdzim opened the pull request:
I wanted the ability to have AIs play against each other locally to compare their strengths. main.go will play all combinations of AIs against each other to determine which is the strongest. I made use of much of the code in pinochle.go. I probably could have used some of the code from server.go to simulate the games, but I wasn't comfortable enough with the code to extract the useful pieces. 

The uint8 and int8s in pinochle.go go frustrating. I was constantly having to cast back and forth and I had some overflows. It simplified the code to just use int everywhere. The go docs said just to use int unless you have a specific reason to use something else and I didn't have a reason, but maybe you did.

This is the basic framework to let AIs play against each other. I plan to start on a strong AI next. Would it be possibly to extract your AI so that I can play against it? I know that I'll at least need to pass a HandTracker to PlayingStrategy.

Feel free to criticize my go style since I'm a go newbie.

bdzim committed:
added the ability to play ais against each other and find the strongest one. Also added some very basic ais

mzimmerman commented:
The uint8's were used to save memory.  Since the AI algorithm I wrote saves the _future_ playing state in memory while it's calculating the best plays, it was running out of memory holding each combination of future states.  Using uint8 reduced the amount of memory required.

In generaly, don't optimize until you have to (use int); that's good practice.

I'll have to re-look at the code myself since it's been a while.  I did implement a json playing API that in theory you could write your own client for that works with the server over the network.

mzimmerman commented on ai/ai.go:
```diff
@@ -0,0 +1,107 @@
+
+func PlayRandom(hand *sdz.Hand, winningCard sdz.Card, leadSuit sdz.Suit, trump sdz.Suit) sdz.Card {
+	for _, v := range rand.Perm(len(*hand)) {
+		if sdz.ValidPlay((*hand)[v], winningCard, leadSuit, hand, trump) {
```
I love your repeated use of sdz.ValidPlay!

mzimmerman commented on pinochle.go:
```diff
@@ -76,7 +76,7 @@ const (
 var Suits [4]Suit
 
-func init() {
+func Init() {
```
init() is a special function in Go - I don't believe Init() will be treated the same.

https://golang.org/doc/effective_go.html#init

mzimmerman commented:
I was able to run your code: MostMeld:PlayLowest is the champ!

Since you're using this as an exercise in learning Go, I'd encourage you to try and make your solution/tests multicore.  See https://golang.org/doc/effective_go.html#concurrency

Your changes did break the tests I had setup on the main package.  Run "go test" inside the main sdzpinochle directory.  There's a dependency upon prettytest that in hindsight, I'd love to get rid of.

mzimmerman commented on test_ai/main.go:
```diff
@@ -0,0 +1,202 @@
+		bidder := playDeal(&match, x%4)
+		pOne := match.Partnerships[0]
+		pTwo := match.Partnerships[1]
+		if pOne.MatchScore > winningScore && pTwo.MatchScore > winningScore {
```
I also see you've taken Uncle John's approach -- must be _over_ 120, but only when it's your opponent that has 120 :)

mzimmerman commented:
In hindsight, I'd also get rid of the dependency on appengine on the server.  It was an interesting idea (make the server instance stateless) but it added a lot of complexity.  Plus, the hardware on appengine was pretty lacking in order to build a proper AI.  As far as I can tell, a good algorithm is expensive in CPU and memory and the free version of Appengine has neither!

bdzim committed:
reverted some changes to remain compatible with server.go

bdzim commented:
Can you try the server tests now? I'm unable to run them locally because I can't import appengine. I reverted all of my changes to pinochle.go so it should work now.

bdzim committed:
add new bidding strategy and track the number of matches played per second

bdzim committed:
test ais in parallel

bdzim commented:
Matches are now simulated in parallel. It went from 370 matches per second to 1470 matches per second with the basic ais I'm using.

mzimmerman committed:
pretty test api update?

mzimmerman commented:
Okay, I see the major disconnect now of why you implemented your own server/game manager.

1) I need to strip out the appengine specific components - as I mentioned, it's not a great server side host for pinochle with a strong AI.  We can even forget the long term storage (or even game handling parts) - Just make it run one table with ~4 clients (some AI, some player, etc)  This will help you with running it locally as it'll remove the barrier to entry.  Go on AppEngine is quite nice, but I understand why you don't want to set it all up (again, it wasn't a good fit anyway)
2) You need to then implement the sdzpinochle/server/Player interface -- just like a human player, my AI needs to see each card that is played, not just the "state" of what it's being asked to play.  The reason it needs to see that is to rule out that the other Ace of spades has already been played.  It also needs to see the meld people lay for the same reason.  Now it knows who has what cards.

mzimmerman commented:
This is the commit I need to go back to - https://github.com/mzimmerman/sdzpinochle/commit/6b079e0e4ca9f8912e910beb54bc768bfd5c7b7f -- Whew!  I hadn't realized how far back my conversion to appengine was!

mzimmerman committed:
removing the appengine core
total hack - no human player support
bumped down strength of AI to  help simulations

mzimmerman commented:
Item 1) from earlier is implemented in 1cb76f4b2f011702eeea91997ab6685ef9cdc419 -- Simply run go test inside server/ -- Before calling NextHand() replace the players with your AI.

bdzim committed:
enabled matches to be simulated by the server

mzimmerman commented:
Closing due to active development being done in noappengine branch
